### ISSUE TITLE
Unable to Deregister Existing Locations in New Catalog Backend

### ISSUE THREAD

**Issue by JP-Dhabolt** [backstage/backstage#6302]
After updating to the new Catalog Backend, any existing locations (registered in the legacy Catalog Backend) cannot be unregistered.  If registering a new component via the new Catalog Backend, the new location can be unregistered without any issues.

## Expected Behavior

Should be able to unregister existing locations after upgrading to the new Catalog Backend.

## Current Behavior

The below popup appears on the screen:
![image](https://user-images.githubusercontent.com/23241077/124127294-8d1eb600-da49-11eb-801e-1f1a6eaffdac.png)

## Steps to Reproduce

1. Have a Component registered via legacy Catalog
2. Update the `catalog.ts` per https://github.com/backstage/backstage/releases/tag/release-2021-06-17
3. Attempt to unregister Component

I am able to reproduce by simply reverting the `catalog.ts` to use the legacy `new CatalogBuilder(env)` pattern and register a new component (no need to revert the library version).  When running in legacy mode, I am able to unregister components created in both legacy and "next" `CatalogBuilder`.

As soon as I update to the `await CatalogBuilder.create(env)` pattern, I receive the error when trying to unregister components created in Legacy, but have no problems unregistering components created in the "next" pattern.

## Context

With our development environment, we are regularly unregistering components and re-adding as we validate various integrations and scenarios.  This slows down our process drastically as we must wait for a catalog refresh cycle to get the latest updates on our component definitions.

In the logs, we see:
```json
{
  "type": "errorHandler",
  "length": 329,
  "name": "error",
  "severity": "ERROR",
  "code": "23503",
  "detail": "Key (id)=(7c68621a-7180-4955-8451-b94b3a6b8f26) is still referenced from table \"entities\".",
  "schema": "public",
  "table": "entities",
  "constraint": "entities_location_id_foreign",
  "file": "ri_triggers.c",
  "line": "3283",
  "routine": "ri_ReportViolation",
  "level": "error",
  "appName": "backstage",
  "environment": "dev",
  "version": "0.1.306",
  "stack": "error: delete from \"locations\" where \"id\" = $1 - update or delete on table \"locations\" violates foreign key constraint \"entities_location_id_foreign\" on table \"entities\"\n    at Parser.parseErrorMessage (/app/node_modules/pg-protocol/dist/parser.js:287:98)\n    at Parser.handlePacket (/app/node_modules/pg-protocol/dist/parser.js:126:29)\n    at Parser.parse (/app/node_modules/pg-protocol/dist/parser.js:39:38)\n    at Socket.<anonymous> (/app/node_modules/pg-protocol/dist/index.js:11:42)\n    at Socket.emit (events.js:315:20)\n    at Socket.EventEmitter.emit (domain.js:467:12)\n    at addChunk (internal/streams/readable.js:309:12)\n    at readableAddChunk (internal/streams/readable.js:284:9)\n    at Socket.Readable.push (internal/streams/readable.js:223:10)\n    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)",
  "message": "delete from \"locations\" where \"id\" = $1 - update or delete on table \"locations\" violates foreign key constraint \"entities_location_id_foreign\" on table \"entities\"",
  "utc_date_time": "2021-07-01T12:41:38.875Z"
}
```

## Your Environment

- NodeJS Version (v12): v14
- Operating System and Version (e.g. Ubuntu 14.04): Docker images `mcr.microsoft.com/vscode/devcontainers/javascript-node:14` and `node:14-alpine` both
- Browser Information: Brave and Chrome

**Comment by freben** [comment]
Hi! Thanks for reporting.

What we did was to try to leave the old data around (the legacy `entities` table) so that there's a downgrade path in the first couple of weeks who may want some time to verify things. What you are seeing is old "ghost" data that references the thing you want to unregister.

We will be looking at adding some interim code to delete related entities in the old table as well, when locations are unregistered. However we just cut a release and the summer vacations are coming up, so I want to help get you unblocked quicker than that.

Your best bet would probably be to either manually drop the `entities_location_id_foreign` foreign key constraint, or to delete all rows out of the `entities` table. Both of those actions are completely safe to do in terms of the _new_ database layout - its active set is in fact in completely different tables. But it does mean committing to the upgrade. Well the second alternative does, the first alternative is relatively easier to recreate, should you ever want to go back. We hope you won't want to, the new system is a lot more scalable and robust!

**Comment by JP-Dhabolt** [comment]
Thanks for the explanation and the workaround!

I will start off by deleting the rows from the `entities` table, since our intention is to continue with the latest and greatest changes.  I will update this issue with the results.

**Comment by JP-Dhabolt** [comment]
I was able to resolve the issue by deleting the rows in the `entities` table.  I solved this by adding a custom DB migration script, which was a bit hacky.  If it is not already on the backlog, supporting custom DB migrations may be something that is useful in the long run.

If anybody else wants to duplicate, here is how I solved this:

```js
// packages/backend/migrations/catalog/20210701110800_delete_legacy_entities.js
// @ts-check

/**
 * @param {import('knex').Knex} knex
 */
exports.up = async function up(knex) {
  return knex.schema.raw('DELETE FROM entities;');
};

exports.down = async function down() {};

```

```ts
// packages/backend/src/plugins/catalog.ts

import {
  CatalogBuilder,
  createRouter,
} from '@backstage/plugin-catalog-backend';
import { Router } from 'express';
import { PluginEnvironment } from '../types';
import { AgeroOrgGithubProcessor } from '../processors';
import { resolvePackagePath } from '@backstage/backend-common';
import fs from 'fs';
import path from 'path';

export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  const builder = await CatalogBuilder.create(env);

  await setupCustomMigrations();

  const {
    entitiesCatalog,
    locationsCatalog,
    locationAnalyzer,
    locationService,
    processingEngine,
  } = await builder.build();

  // Uncomment the below when ready to remove the custom migration.
  // Once this has been deployed, the setupCustomMigrations can be safely removed
  // await removeCustomMigrations(env);

  await processingEngine.start();

  return await createRouter({
    entitiesCatalog,
    locationsCatalog,
    locationAnalyzer,
    locationService,
    logger: env.logger,
    config: env.config,
  });
}

// TODO: Delete this when custom migrations are no longer needed
/*
async function removeCustomMigrations(env: PluginEnvironment) {
  const customMigration = '20210701110800_delete_legacy_entities.js';
  const backstageCatalogMigrations = resolvePackagePath(
    '@backstage/plugin-catalog-backend',
    'migrations',
  );
  const knex = await env.database.getClient();
  await knex.migrate.down({
    name: customMigration,
    directory: backstageCatalogMigrations,
  });
  await fs.promises.unlink(
    path.join(backstageCatalogMigrations, customMigration),
  );
}
*/

async function setupCustomMigrations() {
  const backstageCatalogMigrations = resolvePackagePath(
    '@backstage/plugin-catalog-backend',
    'migrations',
  );
  const customCatalogMigrations = resolvePackagePath(
    'backend',
    'migrations',
    'catalog',
  );

  const migrations = await fs.promises.readdir(customCatalogMigrations);
  const copyPromises = migrations.map(async migration => {
    await fs.promises.copyFile(
      path.join(customCatalogMigrations, migration),
      path.join(backstageCatalogMigrations, migration),
    );
  });
  await Promise.all(copyPromises);
}

```

Finally, don't forget to add `migrations` to the `files` in `packages/backend/package.json`

**Comment by benjdlambert** [comment]
@JP-Dhabolt is it ok to close this?

We will be adding migrations to remove the entities table hopefully in the short term future once the new catalog has settled

**Comment by JP-Dhabolt** [comment]
Yes, I am good with this. Thank you for the assistance!

**Comment by tejaswivinodg** [comment]
👋 @benjdlambert I am seeing the similar issue. I'd like to clarify one more question - if I delete an entity with location, I should be able to re-register the same entity? I see the `register` action says `entity exists` and when I try to refresh the entity it `refresh` endpoint throws 404.
